servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,694 @@
+                            } else if (subscriptionUpdater.compareAndSet(this, prevSubscription, PROCESSING_REQUEST)) {
+                                doRequestMore(prevSubscription);
+                                break;
+                            }
+                        }
+                        return quotaToUse;
+                    }
+                }
+            }
+
+            private void doRequestMore(Subscription prevSubscription) {
+                long availableRequestN = pendingDemandUpdater.getAndSet(this, 0);
+                for (;;) {
+                    if (availableRequestN > 0) {
+                        try {
+                            prevSubscription.request(availableRequestN);
+                        } catch (Throwable cause) {
+                            subscription = CANCELLED;
+                            cancelAndGiveBack(prevSubscription);
+                            throw cause;
+                        }
+                    }
+
+                    final Subscription afterLockSubscription = subscription;
+                    if (afterLockSubscription == PROCESSING_ONNEXT || afterLockSubscription == CANCELLED) {
+                        // onNext is allowed to interrupt, and will handle any pending cancel/request events
+                        // after it completes (otherwise we may drop signals).
+                        break;
+                    } else if (afterLockSubscription == CANCEL_PENDING) {
+                        subscription = CANCELLED;
+                        cancelAndGiveBack(prevSubscription);
+                        break;
+                    } else if (afterLockSubscription == REQUEST_PENDING) {
+                        if (subscriptionUpdater.compareAndSet(this, REQUEST_PENDING, PROCESSING_REQUEST)) {
+                            availableRequestN = pendingDemandUpdater.getAndSet(this, 0);
+                            // continue through next iteration to try to unlock
+                        }
+                    } else if (afterLockSubscription == PROCESSING_REQUEST &&
+                            subscriptionUpdater.compareAndSet(this, PROCESSING_REQUEST, prevSubscription)) {
+                        break;
+                    }
+                }
+            }
+
+            private void cancelAndGiveBack(Subscription subscription) {
+                try {
+                    subscription.cancel();
+                } finally {
+                    giveBackUnusedRequestN();
+                }
+            }
+
+            private boolean giveBackUnusedRequestN() {
+                // we need to give back the outstanding amount that has been requested, but not emitted.
+                // resetting sourceRequestedUpdater must happen unconditionally first, as this stops
+                // calculateSourceRequested from attempting to request any more.
+                final long prevOutstandingDemand = outstandingDemandUpdater.getAndSet(this, -1);
```

Comment:
nit: define a static field for `-1` to make it more clear why this magic number is used.

Reply:
done